### ISSUE TITLE
Adding PS4 controller support

### ISSUE THREAD

**Issue by jimnarey** [Ryzee119/ogx360#44]
Hi,

Thanks for the advice on the rumble issue. I was hoping you might be able to give me some pointers on extending the project?

I have been working on a fork of the project, with the aim of making the hardware build easier and adding support for more controllers. I've removed things like Steel Battalion support and support for multiple controllers (to help me understand what's going on, as much as anything) and it all works fine. I'm using VSCode and the CMake tools extension.

I am now trying to add PS4 controller support:

https://github.com/jimnarey/SimpleXboxControllerAdapter/tree/add_PS4_controller

I started by copying all the files from the main USB Host Library repo which weren't modified in ogx360 into '../lib/USB Host Shield'.

If I add this to main.cpp:

#include <PS4USB.h>

...it builds fine.

If I then add:

PS4USB PS4Wired(&UsbHost); 

I get the following error from the linker:

Scanning dependencies of target sxbca.elf
[ 97%] Building CXX object CMakeFiles/sxbca.elf.dir/main.cpp.obj
[ 98%] Linking CXX executable sxbca.elf
CMakeFiles/sxbca.elf.dir/main.cpp.obj:(.rodata._ZTV9PS4Parser[vtable for PS4Parser]+0x4): undefined reference to `__cxa_pure_virtual'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/sxbca.elf.dir/build.make:103: sxbca.elf] Error 1
make[1]: *** [CMakeFiles/Makefile2:84: CMakeFiles/sxbca.elf.dir/all] Error 2
make: *** [Makefile:84: all] Error 2

If I completely replace the files in '../lib/USB Host Shield' with the all of the files from the main repo, I get exactly the same error.

If I create a new PlatformIO project to test the USB Host library with just:

#include <Arduino.h>
#include <PS4USB.h>

void setup() {
  USB UsbHost;
  PS4USB PS4Wired(&UsbHost);
}

void loop() {
  // put your main code here, to run repeatedly:
}

...it builds fine!

In a separate branch I've started implementing PS3 controller support and while I need to work on translating some of the analog stick readings, it's working as expected and there are no compiler/linker errors.

In Microchip Studio 7, if I move the additional USB Host files into your project (ogx360/Firmware/ogx360/lib/USB Host Shield) I get:
undefined reference to 'HIDUniversal::HIDUniversal(USB*)'
undefined reference to 'PS4Parser::Reset()'
undefined reference to 'PS4Parser::Reset()'

It seems like there's a virtual method which isn't then impemented in a derived class, but this doesn't happen in PlatformIO. I've looked at the CmakeLists file and as far as I can see (I'm new to C++) it's just importing everything in the folder. I also tried running nm -O main.cpp.obj to see if that gave any indication but it lists a load of things as undefined, including digitalWrite! 

         U __cxa_pure_virtual
         U delay
         U digitalRead
         U digitalWrite
00000000 B disconnectTimer
         U __do_clear_bss
         U __do_copy_data
         U __do_global_ctors
         U DukeController_HID_Interface
         U Endpoint_Read_Stream_LE
00000000 B enumerationComplete
00000000 t _GLOBAL__sub_I_XboxOGDuke
         U HID_Device_USBTask
         U init
00000000 T main
         U millis
         U pinMode
00000000 B PS4Wired
         U SetupHardware
0000003e a __SP_H__
0000003d a __SP_L__
0000003f a __SREG__
00000000 a __tmp_reg__
00000000 B UsbHost
         U USB_USBTask
00000000 B Xbox360Wired
00000000 B XboxOGDuke
00000000 B XboxOneWired
00000000 T setRumbleOn(unsigned char, unsigned char)
00000000 T getAnalogHat(AnalogHatEnum)
00000000 T getButtonPress(ButtonEnum)
00000000 T controllerConnected()
00000000 T sendControllerHIDReport()
00000000 T setLedOn(LEDEnum)
00000001 a __zero_reg__
00000000 W HIDUniversal::GetAddress()
         U HIDUniversal::EndpointXtract(unsigned char, unsigned char, unsigned char, unsigned char, USB_ENDPOINT_DESCRIPTOR const*)
         U HIDUniversal::GetReportParser(unsigned char)
         U HIDUniversal::SetReportParser(unsigned char, HIDReportParser*)
         U HIDUniversal::Init(unsigned char, unsigned char, bool)
         U HIDUniversal::Poll()
00000000 W HIDUniversal::isReady()
         U HIDUniversal::Release()
         U HIDUniversal::HIDUniversal(USB*)
00000000 W USBDeviceConfig::DEVCLASSOK(unsigned char)
00000000 W USBDeviceConfig::ResetHubPort(unsigned char)
00000000 W USBDeviceConfig::DEVSUBCLASSOK(unsigned char)
00000000 W USBDeviceConfig::ConfigureDevice(unsigned char, unsigned char, bool)
         U USB::outTransfer(unsigned char, unsigned char, unsigned int, unsigned char*)
         U USB::Task()
         U USB::USB()
00000000 W TPin<Portb, (unsigned char)6>::Set()
00000000 W PS4USB::ParseHIDData(USBHID*, bool, unsigned char, unsigned char*)
00000000 W PS4USB::OnInitSuccessful()
00000000 W PS4USB::sendOutputReport(PS4Output*)
00000000 W PS4USB::VIDPIDOK(unsigned int, unsigned int)
         U XBOXONE::getAnalogHat(AnalogHatEnum)
         U XBOXONE::getButtonPress(ButtonEnum)
         U XBOXONE::XBOXONE(USB*)
         U XBOXUSB::getAnalogHat(AnalogHatEnum)
         U XBOXUSB::getButtonPress(ButtonEnum)
         U XBOXUSB::setLedOn(LEDEnum)
         U XBOXUSB::XBOXUSB(USB*)
00000000 W MAX3421e<TPin<Portb, (unsigned char)6>, TPin<Portb, (unsigned char)5> >::Init()
00000000 t MAX3421e<TPin<Portb, (unsigned char)6>, TPin<Portb, (unsigned char)5> >::regRd(unsigned char) [clone .isra.5]
00000000 t MAX3421e<TPin<Portb, (unsigned char)6>, TPin<Portb, (unsigned char)5> >::regWr(unsigned char, unsigned char) [clone .isra.6]
00000000 W MAX3421e<TPin<Portb, (unsigned char)6>, TPin<Portb, (unsigned char)5> >::busprobe()
00000000 V MAX3421e<TPin<Portb, (unsigned char)6>, TPin<Portb, (unsigned char)5> >::vbusState
         U PS4Parser::Parse(unsigned char, unsigned char*)
         U PS4Parser::Reset()
0000007e W non-virtual thunk to PS4USB::sendOutputReport(PS4Output*)
         U non-virtual thunk to HIDUniversal::EndpointXtract(unsigned char, unsigned char, unsigned char, unsigned char, USB_ENDPOINT_DESCRIPTOR const*)
00000000 V vtable for PS4USB
00000000 V vtable for PS4Parser
00000000 b main::commandTimer
00000000 b main::xboxHoldTimer
00000000 b main::report

I'm not quite sure where to start? Any help would be greatly appreciated!

Thanks, Jim.

**Comment by jimnarey** [comment]
Hi - I get similar errors, e.g.

libUSB_Host_Shield.a(Wii.cpp.obj):(.rodata._ZTV16BluetoothService[vtable for BluetoothService]+0x4): undefined reference to `__cxa_pure_virtual'

...if I try to instantiate the PS4BT or WII classes. I can instantiate the general hiduniversal class. I now have the PS3 controller completely working.

**Comment by Ryzee119** [comment]
The __cxa_pure_virtual function is defined in `abi.cpp` as part of the arduinocore. This file has not been copied over to this project.
Hopefully copying this https://github.com/arduino/ArduinoCore-avr/blob/master/cores/arduino/abi.cpp into the `Firmware/ogx360_32u4/lib/Arduino/` folder should do the business :)

**Comment by jimnarey** [comment]
Hi - thank you so much for the help. I had thought `__cxa_pure_virtual' was a name produced by the compiler and didn't realise I could have searched for it within the Arduino library. So I'll know for next time.

I tried copying abi.cpp but got errors from the compiler this time, about the contents of the formerly missing function:

https://github.com/jimnarey/SimpleXboxControllerAdapter/blob/test_lib_imports/add_abi_dot_cpp_error

..so I tried copying all of the files in ArduinoCore/cores/arduino into the project and got a different set of errors compiling USBCore, seemingly before it gets to abi.cpp ("‘USB_VID’ was not declared in this scope"):

https://github.com/jimnarey/SimpleXboxControllerAdapter/blob/test_lib_imports/add_all_missing_Arduino_files_error

If I delete the contents of '../lib/Arduino' in the project altogether, copy everything from 'ArduinoCore/cores/arduino' in the Arduino library into it and then replace the missing pins_arduino.h from the Arduino library's 'variants' folder, I get the second set of errors again.

Any more help with this would be really appreciated - I'm really keen to get some more controllers working! Thank you!

**Comment by jimnarey** [comment]
Hi - I've sorted it, albeit not in a very clean way. The first error thrown when only abi.cpp is put in the Arduino folder is:

/home/jimnarey/projects/SimpleXboxControllerAdapter/Firmware/lib/Arduino/abi.cpp:25:3: error: expected unqualified-id before ‘[’ token
   [[gnu::weak, noreturn]] void terminate() {
   ^

So I tried removing "[[gnu::weak, noreturn]]" and it compiled. I get a couple of compiler warnings about no return functions returning, but everything works fine and I have just added in PS4 support via USB.

Thank you so much for your help and all the work that went into ogx360!
